taichi/llvm/llvm_program.cpp

```diff
@@ -564,5 +567,26 @@ DevicePtr LlvmProgramImpl::get_snode_tree_device_ptr(int tree_id) {
   return tree_alloc.get_ptr();
 }
 
+DeviceAllocation LlvmProgramImpl::allocate_memory_ndarray(
+    std::size_t alloc_size) {
+  Device::AllocParams device_buffer_alloc_params;
+  device_buffer_alloc_params.size = alloc_size;
+  DeviceAllocation alloc =
+      get_compute_device()->allocate_memory(device_buffer_alloc_params);
+  ndarray_allocs_.push_back(alloc);
+  return alloc;
+}
+
+uint64_t *LlvmProgramImpl::get_ndarray_alloc_info_ptr(DeviceAllocation &alloc) {
+  if (config->arch == Arch::cuda) {
+#if defined(TI_WITH_CUDA)
+    return (uint64_t *)cuda_device()->get_alloc_info(alloc).ptr;
+#else
+    TI_NOT_IMPLEMENTED
```

Comment:
I'm always curious if this if-else is necessary :P When do we have arch== cuda but TI_WITH_CUDA not enabled? (In this case I thought we already fallback to cpu?

Reply:
when someone wrote ti.init(arch=cuda) but build Taichi with TAICHI_CMAKE_ARGS="-DTI_WITH_CUDA:BOOL=OFF"? ;)))

Reply:
In practice this never happens since we build our wheels using TI_WITH_CUDA=ON? hmmm I'm just a bit confused :P

Reply:
(fyi we figured out this macro is indeed turned off on mac so nvm :D